hail/src/main/scala/is/hail/expr/types/physical/PArray.scala

```diff
@@ -7,25 +7,28 @@ import is.hail.expr.ir.EmitMethodBuilder
 import is.hail.expr.types.virtual.TArray
 import is.hail.utils._
 import scala.reflect.{ClassTag, _}
+object PArray {
+  def apply(elementType: PType, required: Boolean = false) = new PCanonicalArray(elementType, required)
```

Comment:
1) allows us to gradually convert PArray( ) constructors to PCanonicalArray.
2) One possible way of deciding which implementation of PArray to return (canonical or non-canonical), is by the IR. Another possibility is to have this class decide that, based on some properties of the element type (such as sparsity)

Reply:
yes, this is a nice way to avoid having to refactor a bunch of other code.